File: Confused Number.py
# Time:- O(2^n)
# Space:-O(2^n)
# Approach:- We have some options on which we can backtrack and form a number and then we can check if
# the number formed is less than the given number and if we reverse it does it form the same number of  or not, if yes then we can increase the result count.
class Solution:
    def confusingNumberII(self, N: int) -> int:
        options=[0,1,6,8,9]
        self.res=0
        # reverse the number and check if it is same as the old number
        def valid(number):
            maps=[0,1,0,0,0,0,9,0,8,6]
            newnumber=0
            oldnumber=number
            while(number>0):
                newnumber=newnumber*10+maps[number%10]
                number//=10
            if newnumber==oldnumber:
                return False
            return True
            
        def cn(number):
            # if number formed is greater than n return
            if number>N:
                return
            # if number formed is less than n and valid then increase res count
            if number<=N and valid(number):
                self.res+=1
            # for all the number in options backtrack and try to form the numbers
            for i in options:
                # action
                number=number*10+i
                # number can't be 0
                if number==0:
                    continue
                # recurse
                cn(number)
                # backtrack
                number//=10
        cn(0)
        return self.res

sol=Solution()
sol.confusingNumberII(100)

File: confusedNumber.java
// Time Complexity = Exponential / O(5^n)
// Space Complexity = O(L), where L=no. of digits present in n
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
class Solution {
    int count;
    HashMap<Integer, Integer> map;

    public int confusingNumberII(int n) {
        if (n==0) return 0;
        map = new HashMap<>();
        map.put(0, 0);
        map.put(1, 1);
        map.put(6, 9);
        map.put(8, 8);
        map.put(9, 6);

        dfs(n, 0);
        return count;
    }

    // We are creating more numbers using the valid numbers in the map till we reach n
    private void dfs(int n, long curr) {
        // base
        if (curr > n) return;

        // logic
        if (isValid(curr)) count++;

        for (int key: map.keySet()) {
            long newNumber = (curr * 10) + key;
            if (newNumber != 0) {
                dfs(n, newNumber);
            }
        }
    }

    // This function is rotating the number in 180 degree and checking if its valid by comparing with the original number and if they are different then its valid
    private boolean isValid (long curr) {
        long result = 0;
        long original = curr;

        while(curr > 0) {
            result = (result*10)+map.get((int)(curr%10));
            curr = curr/10;
        }
        return original != result;
    }
}